doc/cpp_simulation.rst

```diff
@@ -0,0 +1,139 @@
+        * **Experimental inputs** that can change between model runs, such
+          as external spike trains.
+
+    Simulations provide an interface for executing and interacting with the model:
+
+        * **Advance model state** from one time to another and reset model
+          state to its original state before simulation was started.
+        * **I/O** interface for sampling simulation variables (e.g. compartment voltage
+          and current) and spike output.
+
+    **Types:**
+
+    .. cpp:type:: communicator_type = communication::communicator<communication::global_policy>
+
+        Type used for distributed communication of spikes and global synchronization.
+
+    .. cpp:type:: spike_export_function = std::function<void(const std::vector<spike>&)>
+
+        User-supplied callack function used as a sink for spikes generated
+        during a simulation. See :cpp:func:`set_local_spike_callback` and
+        :cpp:func:`set_global_spike_callback`.
+
+    **Constructor:**
+
+    .. cpp:function:: simulation(const recipe& rec, const domain_decomposition& decomp)
+
+    **Experimental inputs:**
+
+    .. cpp:function:: void inject_events(const pse_vector& events)
+
+        Add events directly to targets.
+        Must be called before calling :cpp:func:`simulation::run`, and must contain events that
+        are to be delivered at or after the current simulation time.
+
+    **Updating Model State:**
+
+    .. cpp:function:: void reset()
+
+        Reset the state of the simulation to its original state before
+        :cpp:func:`simulation::run` was called.
+
+    .. cpp:function:: time_type run(time_type tfinal, time_type dt)
+
+        Run the simulation from current simulation time to :cpp:var:`tfinal`,
+        with maximum time step size :cpp:var:`dt`.
+
+    .. cpp:function:: void set_binning_policy(binning_kind policy, time_type bin_interval)
+
+        Set event binning policy on all our groups.
+
+    **I/O:**
+
+    .. cpp:function:: sampler_association_handle add_sampler(\
+                        cell_member_predicate probe_ids,\
+                        schedule sched,\
+                        sampler_function f,\
+                        sampling_policy policy = sampling_policy::lax)
+
+        Note: sampler functions may be invoked from a different thread than that
+        which called :cpp:func:`simulation::run`.
+
+        (see the :ref:`sampling_api` documentation.)
+
+    .. cpp:function:: void remove_sampler(sampler_association_handle)
+
+        Remove a sampler.
+        (see the :ref:`sampling_api` documentation.)
+
+    .. cpp:function:: void remove_all_samplers()
+
+        Remove all samplers from probes.
+        (see the :ref:`sampling_api` documentation.)
+
+    .. cpp:function:: std::size_t num_spikes() const
+
+        The total number of spikes that occurred since either construction or
+        last resetting the model.
+
```

Comment:
Is this spikes generated, or delivered? I.e. how are spikes 'in flight' counted with respect to this number? We should also clarify that this is a global (across all domains) count.

Reply:
I thought about whether to say "global" model. I think that the "global" is redundant: "the model" really means "the model".

I will update to clarify that it is generated spikes.